accounting/invoice/utils.py

```diff
@@ -21,27 +21,102 @@
 Invoice utilities.
 """
 
-from datetime import timedelta
+import datetime
 
-from django.utils import timezone
+from django.conf import settings
+from django.core.mail.message import EmailMessage
+from django.template.loader import get_template
 
+from accounting.account.models import Account
+from accounting.invoice import choices, models
 
-def get_last_day_past_month():
+
+def resolve_dict_callables(dictionary, *args):
     """
-    Return the `datetime` object representing the last day of the month that just passed.
+    Resolve any potential callables in a dictionary by passing it some arguments.
     """
-    return timezone.now().replace(day=1) - timedelta(days=1)
+    # We use a new dictionary to prevent mutating the callable. If we didn't, then on another round of calling this
+    # with the same dictionary, the callable would have disappeared.
+    new_dict = {}
+    for key, potential_callable in dictionary.items():
+        if callable(potential_callable):
+            new_dict[key] = potential_callable(*args)
+        else:
+            new_dict[key] = dictionary[key]
+    return new_dict
 
 
-def get_first_day_past_month():
-    """
-    Return the `datetime` object representing the first day of the month that just passed.
+# pylint: disable=too-many-arguments,too-many-locals
+def send_email_with_invoice(template, subject, extra_email_context=None, client_usernames=settings.BILLING_CYCLE_USERS,
+                            create_invoice=False, draft_invoice=False, fill_line_items_from_jira=False,
+                            upload_to_google_drive=False, auto_approve=False):
     """
-    return get_last_day_past_month().replace(day=1)
+    Send an email with an invoice attached, and potentially perform peripheral actions.
 
+    TODO: This function's pretty beefy; break it up by encapsulating logic into more utility functions or a new class.
 
-def default_invoice_number():
-    """
-    Return the invoice number which defaults to `yyyy-mm` for the current year and month.
+    :param template: The email template.
+    :param subject: The email subject.
+    :param extra_email_context: Extra kwargs to update the email context with.
+    :param client_usernames: The list of billing cycle users' usernames.
+    :param create_invoice: Whether to create a new invoice based off of a past one.
+    :param draft_invoice: Whether this is a draft invoice or not.
+    :param fill_line_items_from_jira: Whether to fill up the invoice's line items from JIRA worklogs.
+    :param upload_to_google_drive: Whether to upload the resulting invoice PDF to Google Drive.
```

Comment:
@UmanShahzad Could you extend this to describe the purpose of "draft invoices", where this code will store them, etc.?

Reply:
Yes, sure :+1: